test-js/test-utils.js

```diff
@@ -0,0 +1,247 @@
+    expect(numbersreversed.next().value).equals(2);
+    expect(numbersreversed.next().value).equals(1);
+  });
+});
+
+describe("enumerate", () => {
+  it("enumerate an array, enumeration start with 0 by default", () => {
+    const numbers = [1, 2, 3];
+    const numbersenumerated = enumerate(numbers);
+    expect(numbersenumerated.next().value[0]).equals(0);
+    expect(numbersenumerated.next().value[0]).equals(1);
+    expect(numbersenumerated.next().value[0]).equals(2);
+  });
+  it("enumaration start with a different number than 0", () => {
+    const numbers = [1, 2, 3];
+    const numbersenumerated = enumerate(numbers, 1);
+    expect(numbersenumerated.next().value[0]).equals(1);
+    expect(numbersenumerated.next().value[0]).equals(2);
+    expect(numbersenumerated.next().value[0]).equals(3);
+  });
+});
+
+describe("reversedEnumerate", () => {
+  it("enumerate and reverse an iterator", () => {
+    const numbers = [1, 2, 3];
+    const numbersreversed = reversedEnumerate(numbers);
+    {
+      const iteratorResult = numbersreversed.next();
+      expect(iteratorResult.value[0]).equals(2);
+      expect(iteratorResult.value[1]).equals(3);
+    }
+    {
+      const iteratorResult = numbersreversed.next();
+      expect(iteratorResult.value[0]).equals(1);
+      expect(iteratorResult.value[1]).equals(2);
+    }
+    {
+      const iteratorResult = numbersreversed.next();
+      expect(iteratorResult.value[0]).equals(0);
+      expect(iteratorResult.value[1]).equals(1);
+    }
+  });
+});
+
+describe("range", () => {
+  it("should generate a range of numbers", () => {
+    const numbers = range(3);
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(0);
+    }
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(1);
+    }
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(2);
+    }
+  });
+  it("should generate a range of numbers, with start and stop values", () => {
+    const numbers = range(10, 13);
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(10);
+    }
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(11);
+    }
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(12);
+    }
+  });
+  it("should generate a range of numbers, with start, stop and step values", () => {
+    const numbers = range(10, 15, 2);
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(10);
+    }
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(12);
+    }
+    {
+      const iteratorResult = numbers.next();
+      expect(iteratorResult.value).equals(14);
+    }
+  });
+});
+
+describe("chain", () => {
+  it("chain iterators", () => {
+    function* numbers() {
```

Comment:
You could also use `range()` here, as that returns an iterator, too.

Reply:
But at this case, if the range function fails (we have different test cases for it), this will fail too. So the chain test will be dependent to range test case. What do you think?